spec/Section 2 -- Language.md

```diff
@@ -980,3 +985,292 @@ and operations.
+
+A GraphQL Document which contains {TypeSystemDefinition} must not be executed;
+GraphQL execution services which receive a GraphQL Document containing type
+system definitions should return a descriptive error.
+
+Note: This IDL is used throughout the remainder of this specification document
+when illustrating example type systems.
+
+
+### Schema Definition
+
+SchemaDefinition : schema { OperationTypeDefinition+ }
+
+OperationTypeDefinition : OperationType : NamedType
+
+A GraphQL Type System includes at most one Schema Definition, which defines
+the *root types* to be used for each operation.
+
+In this example, a GraphQL schema is defined with both query and mutation
+root types:
+
+```graphql
+schema {
+  query: MyQueryRootType
+  mutation: MyMutationRootType
+}
+
+type MyQueryRootType {
+  someField: String
+}
+
+type MyMutationRootType {
+  setSomeField(to: String): String
+}
+```
+
+**Default Root Types**
+
+While any type can be the *root type* for a GraphQL query or mutation operation,
+GraphQL type system definitions can omit the schema definition when the query
+and mutation root types are named `Query` and `Mutation`, respectively.
+
+Similarly, when serializing a GraphQL schema using the type system language, a
+schema definition should be omitted if only uses the default root type names.
+
+This example describes a valid complete GraphQL schema, despite not explicitly
+including a schema definition. The `Query` type is presumed to be the query
+root type of the schema.
+
+```graphql
+type Query {
+  someField: String
+}
+```
+
+
+### Type Definition
+
+TypeDefinition :
+  - ScalarTypeDefinition
+  - ObjectTypeDefinition
+  - InterfaceTypeDefinition
+  - UnionTypeDefinition
+  - EnumTypeDefinition
+  - InputObjectTypeDefinition
+
+A GraphQL Type System is defined by many different kinds of types.
+
+
+#### Scalar
+
+ScalarTypeDefinition : scalar Name
+
+Scalar types represent leaf values in a GraphQL type system. While this GraphQL
+specification describes a set of Scalar types which all GraphQL services must
+supply, custom Scalar types may be supplied by a GraphQL service. Typically, the
+set of Scalar types described in this specification are omitted from documents
+which describe a schema for brevity.
+
+In this example, a Scalar type called `DateTime` is defined:
+
+```graphql
+scalar DateTime
+```
+
+
+#### Object
+
+ObjectTypeDefinition : type Name ImplementsInterfaces? Directives? { FieldDefinition+ }
+
+ImplementsInterfaces : implements NamedType+
+
+FieldDefinition : Name ArgumentsDefinition? : Type Directives?
+
+ArgumentsDefinition : ( InputValueDefinition+ )
+
+InputValueDefinition : Name : Type DefaultValue? Directives?
+
+Object types represent a list of named fields, each of which yield a value of a
+specific type. Each field itself may accept a list of named arguments.
+
+Objects may implement Interface types. When implementing an Interface type, the
+Object type must supply all fields defined by the Interface.
```

Comment:
I really like, that interfaces exists in GraphQL. But as a computer scientist this feels unnatural for me.
Besides the feeling I can think of at least the following problem:
1. Alice defines an Interface A with Field x in her Project.
2. Bob adds the project of Alice as dependency and `implements` her Interface A in type B of his project.
3. Alice adds Field y to Interface A and publishes her changes.
4. Bob updates his dependencies -> The GraphQL implementation he uses reports an error.

This currently means, that adding a field to an interface is a major change and thus not backwards compatible.
We would get around this problem, if `the object type MAY supply the fields defined by the interface`.

My feeling is, that forcing the user of an Interface to repeat the same fields, does not add value while reducing maintainability.

What is the reason that one has to repeat all fields that are defined in the interface? 

Kind regards :)

Reply:
Surely that's literally the point of an interface, that those who "implement" it are actually doing so.

Reply:
You agree with you about the definition of an interface and that the current usage of the `implements` keyword in GraphQL matches the definition. I was not able to correctly describe the issue.

So here is what I mean:
Let us define, that types that implement an interface automatically satisfy it. This does not require any Syntax addition or change. We only have to change one word in the languager definition (MUST-> MAY) above and do not need additions like mixins (#370).
There is at least one other person (@SerafimArts) suggesting this (https://github.com/facebook/graphql/issues/370#issuecomment-339848304).

In my view this is the simplest change in the spec that would at the same time improve the productivity of many people because they do not have to repeat the fields of the interface over and over again.

What speaks against this?

Reply:
> What speaks against this?

The fact that this would render interfaces completely meaningless?

Reply:
Seems like it is still not clear what I mean. 
I suggest, that an type that implements an interface implicitly has all the fields of the interface.

Reply:
@mohawk2 
>The fact that this would render interfaces completely meaningless?

Why is that? The interface defines those fields that the type (object) is obliged to contain. But this does not mean that we have to copy them from the interface to the implementation.

Interfaces in the modern world are needed in order to oblige classes to implement methods. 
In GraphQL/SDL, there are no methods, no realizations, and this turns out to be stupid Ctrl+C -> Ctrl+V.

### e.g Current behavior

```graphql
interface B {
    field: Any
}

# Type A with field "field"
type A implements B { 
    field: Any # Copy+Paste
}

 # Error: Field "field" required
type C implements B {
}
```

### e.g Proposal behavior

```
interface B {
    field: Any
}

# Type A with field "field"
type A implements B {
    field: Any # Copy+Paste
}

# Type C with field "field"
type C implements B {
    # Auto resolving. We pretend that the definition in the Object completely repeats the definition in the Interface.
}
```

Reply:
I would suggest this means you should generate your schemas rather than manually maintaining them?

Reply:
@mohawk2 Schemes of introspection and so generated... Or what is the question? Generation of SDL? How does a little tuning of the gql standard and its implementation?

Reply:
@mohawk2 Can you explain what you mean? I also don't understand your sentence :)

Reply:
I am trying to indicate that I believe you are solving the wrong problem, and that the way to avoid incorrectness in IDL representation of implementers of interfaces is to automatically generate those IDL representations. In any case, I do not thing the IDL should change in the way you suggest.

Reply:
Generate SDL? :D Lol?

Reply:
To close the loop on this conversation, we've now discussed this thoroughly. This is a tradeoff between explicitness (requiring object definitions to declare all fields they implement) and terseness (imply an object contains the fields implements in an interface).

There are a number of reasons why we've chosen to lean in favor of explicitness in this instance:

* Readability: The SDL is designed to be read far more often than it is written (in fact, in many cases it is "written" by a code generation script). Because of this, it's highly desirable to see all defined fields in one place and not have to follow references to interfaces to understand the full set of fields defined.

* Removal of adherence ambiguity: Consider the following example schema with implicitly defined fields:

  ```graphql
  interface Foo {
    self: Foo
  }
  interface Bar {
    self: Bar
  }
  type FooBar implements Foo, Bar {}
  ```

  In this case, what is the type of the field `self`? While this sort of ambiguity can be solved by additional checks and constraints, it complicates tools and can lead to pitfalls when designing APIs. Requiring fields to be defined removes this complication and potential for ambiguity.